### DISCUSSION TITLE
Version 0.6

### DISCUSSION THREAD

**Discussion by ililim** [ililim/dual-key-remap#36] [Announcements]
Download [dual-key-remap.zip](https://github.com/ililim/dual-key-remap/releases/download/v0.6/dual-key-remap-v0.6.zip) to install. Installation instructions can be found in the [README](https://github.com/ililim/dual-key-remap#installation).

### Changed
- Improve how we send inputs, Dual-key-remap should now work in more applications and locales!
- Removed virtual keycodes as all key inputs are now sent as virtual keycodes. If you relied on this previously to make the app work you can use the regular keycodes again.
- Dual key remap will no longer try to automatically remap both modifiers keys if you don't specify whether it's the left or the right key. If you don't provide a side then it will rebind the LEFT key by default. For those that need both keys remapped, you can create two remappings in your config.
- Renamed and cleaned up the available key codes. You may have to update your config.
- If a keycode that you need is missing open a new issue to report it.
### Fixed
- Remappings bound to left or right modifiers will no longer be triggered by both left and right keys.

<hr /><em>This discussion was created from the release <a href='https://github.com/ililim/dual-key-remap/releases/tag/v0.6'>Version 0.6</a>.</em>

**Comment by dumblob** [comment]
That's great news, thanks! I have some questions if you don't mind.

1) Does this everything work also on Windows 7? If not, what's the minimum version and/or SW equipment needed to successfully use this?

2) Do you think I'd be able to configure all the following simultaneously (assuming default US keyboard layout for clarity)?

pressed in this order | resulting dual-key-remap action | discards all the original pressed keys so that they won't reach the app or just passthrough
-|-|-
a | no action (i.e. letter `a`) | passthrough
LShift + a | no action (i.e. letter `A`) | passthrough
RShift + a | no action (i.e. letter `A`) | passthrough
LShift + RShift + a | glyph `&` | discard
RShift + LShift + a | glyph `&` | discard
LShift#press + RShift#press + LShift#depress + a | glyph `&` | discard
RShift#press + LShift#press + RShift#depress + a | glyph `&` | discard
LShift + a + RShift | no action | passthrough
a + LShift + RShift | no action | passthrough
RShift + a + LShift | no action | passthrough
a + RShift + LShift | no action | passthrough
a + LShift + RShift | no action | passthrough
LShift + RShift | no action | passthrough
RShift + LShift | no action | passthrough

(no `#...` suffix means "press & hold" otherwise the suffix specifies it)

**Reply by ililim** [reply]
I'm happy to answer you question but have to ask - remapping two shifts + a -> `&` is quite the peculiar remapping. Why are you looking for this one in particular?

1. It should work on win7 I believe - but I haven't tested it myself.
2. Dual Key Remap is designed to remap one modifier to two, and if I'm reading your table right it looks like you're remapping three keys to one (both shifts + a = &). For general purpose remapping you'd be better off writing a script in AutoHotkey which should be capable of achieving the above with a bit of effort.

**Reply by dumblob** [reply]
> I'm happy to answer you question but have to ask - remapping two shifts + a -> `&` is quite the peculiar remapping. Why are you looking for this one in particular?

Some 10 years ago I designed a layout which I use on every computer I come to. It's a standard US layout (it's dvorak, but that doesn't matter at all) but with completely deleted 3rd and any higher layer. And instead my own 3rd layer added. The key feature of this my layout is, that to get the 3rd level you have to press both shifts at the same time (and optionally release either of them). The `&` is just one example from the 3rd layer.

This is impossible to do with conventional layouting on any Windows because it seems since Win98/Win2000 the low-level code of keyboard handling (there is some leaked source code which clearly shows that - I can dig it out if you want) maps LShift to "Shift" and RShift also to "Shift". It's the only hard-coded mapping which exists in the low-level code (based on the examination of the leaked source code). And probably due to backwards compatibility it's impossible to change this choice.

On macOS, Linux/X11, Linux/Wayland, etc. this poses, of course, absolutely no problem at all.

> 1. It should work on win7 I believe - but I haven't tested it myself.

Thanks, that's great! If not as is, then for study purposes.

> 2. Dual Key Remap is designed to remap one modifier to two, and if I'm reading your table right it looks like you're remapping three keys to one (both shifts + a = &).

Yes, that's correct.

>For general purpose remapping you'd be better off writing a script in AutoHotkey which should be capable of achieving the above with a bit of effort.

AHK had some issues with missed key presses, timing issues, and weird corner cases and missing support in some applications (e.g. M$ Office is notorious for not working well with custom mappings). That was about 5 year ago. Maybe it's different now or maybe (quite probably) I just didn't have the patience to debug all those tens of issues back then.

Feel free to tell me what'd be better - whether extending dual-key-remap or forking it and adjusting to the 3rd level or yet something else... IDK

**Reply by ililim** [reply]
Interesting! Makes sense, that's a neat approach might give that a shot one of these days :)

Honestly if I were to recommend something I'd suggest getting one of those programmable mechanical keyboards. I believe there is some out there that support such layer functionality explicitly - and they would work regardless of where they are plugged in.

As for software solutions, AHK isn't perfect but it's probably the most time efficient to work with. If you use I'd recommend remapping using virtual keys - that way you can distinguish shift keys; and to write a custom AHK script as opposed going with their default remapping syntax (so sending input using their SendInput function, managing states yourself with variables and conditionals).

Finally there is the option of writing your own c utility. Probably the least practical approach; and it's also the solution I took (that's why Dual Key Remap exists 😉). Forking this codebase is probably a good place to start if you're comfortable writing or tinkering with c code. If you do I'd recommend ripping out the config/remap functionality and to just hardcode in your usecase in `handle_input`. For instance a starting point could be:

```c
int l_shift_state = UP;
int r_shift_state = UP;

/* @return swallow_input */
int handle_input(int scan_code, int virt_code, int direction, int is_injected)
{
    if (virt_code == VK_LEFT_SHIFT) {
        l_shift_state = direction;
    }
    if (virt_code == VK_RIGHT_SHIFT) {
        r_shift_state = direction;
    }

    int third_layer_active = l_shift_state == DOWN && r_shift_state == DOWN;

    if (direction == DOWN && third_layer_active) {
        switch (virt_code)
        {
        case VK_KEY_A:
            // This is prototype code so it doesn't handle edge cases :)
            //
            // For instance: it immidiately sends both UP and DOWN. Ideally we'd
            // only send DOWN now and wait for the key to lift before sending UP.

            // In this basic implementation key repeat is going to be very slow
            // as we'll be basically sending DOWN UP DOWN UP DOWN UP over and over.

            // Send `&` DOWN UP
            // Since shift is already down we can send '7' to get '&'
            // If we wanted to send a non-shifted key we'd have to depress
            // shift and handle that correctly in the `third_layer_active` flag.
            send_input(0, VK_KEY_7, DOWN);
            send_input(0, VK_KEY_7, UP);
            break;

        default:
            break;
        }
    }

    // Swallow non-injected keys while third layer active
    if (!is_injected && third_layer_active) {
        return 1;
    }
    // Otherwise passthrough
    return 0;
}
```

Just a start - other than listed limitations I don't think this quite has your desired behavior of optional release of one of the shifts in there. Writing it yourself is definitely the most work but you'd be able to get exactly what you need.

**Reply by dumblob** [reply]
> Interesting! Makes sense, that's a neat approach might give that a shot one of these days :)

Sure, go ahead! I have some rudimentary skeletons in https://github.com/dumblob/ULKL (some working, some not). The "double shift trick" proved very efficient over time (at least to me :wink:) - it allows one to use muscle memore because your brain can tell the correct hand and the given finger:

>hey, you'll be pressing a letter key - but first let your pinkie **always on the very same hand** press Shift, then in the meantime the other hand will also press & hold Shift and you can easily release the first hand Shift and finally press the desired key

It feels so intuitive I didn't need to learn anything - it's the same hand as the final desired key. The whole ULKL concept is built around "not learning" :wink: - once you learned US layout, you just intuitively use the same finger to press the "modification" (e.g. German `ä` is written with the same finger as `a`, German `ö` is written with the same finger as `o` etc. - same for Finnish, French, Slovak, Czech, Polish etc. - but I didn't finish all these layouts; of course there are some exceptions for special characters, but they're so extremely rare and uncommon in normal text you won't mind remembering them because this tiny learning process works on its own subconsciously). Ok, end of marketing :wink:.

> Honestly if I were to recommend something I'd suggest getting one of those programmable mechanical keyboards. I believe there is some out there that support such layer functionality explicitly - and they would work regardless of where they are plugged in.

That's an interesting thought. How would you circumvent that HID keyboards all go through this hardcoded LShift/RShift->Shift mapping? I can imagine custom USB device driver which will deliberately not be HID and which will then generate key presses in a way it doesn't go through the hardcoded mapping (maybe that path will be avoided if virtual keys will be generated - didn't test it yet - that's why I'm asking here :wink:). But that's then pretty much the same as "do it in SW" :wink:.

> As for software solutions, AHK isn't perfect but it's probably the most time efficient to work with. If you use I'd recommend remapping using virtual keys - that way you can distinguish shift keys; and to write a custom AHK script as opposed going with their default remapping syntax (so sending input using their SendInput function, managing states yourself with variables and conditionals).

This sounds really important, thanks!

> Finally there is the option of writing your own c utility. Probably the least practical approach; and it's also the solution I took (that's why Dual Key Remap exists ). Forking this codebase is probably a good place to start if you're comfortable writing or tinkering with c code. If you do I'd recommend ripping out the config/remap functionality and to just hardcode in your usecase in `handle_input`. For instance a starting point could be:

Thanks a lot! Thinking about this further, I'm not sure AHK is allowed by all antivirus SW in companies etc. So if I want the solution to work really everywhere, I'm still probably better off with my own public C code and a binary. I'm currently without any Windows machine, but in couple of weeks I suppose I'll need to use a few again. So if I won't find any solution until then, I'll consider writing it on those machines.

**Reply by dumblob** [reply]
So it seems it's pretty difficult to make such an app work easily on a login screen. The best approach seems to be to write a credential provider but except for more work & maintenance this requires the problem might be the deployment but especially the user experience when (according to my shallow understanding) a user has to each time choose the correct credential provider from a list under the input form. That's super inconvenient.

So I just came up with yet another potential hacky approach - namely to use custom written IME for this (invisible, doing just the RShift- & RShift- specific mappings and no more; should be just one tiny DLL) and leaving the rest to some standard keyboard layout. I've never worked with IME, so I can't tell whether it's as seamless as it sounds nor whether this'll work painlessly "by default" on a login screen.

Thoughts?

**Reply by ililim** [reply]
Interesting! Feel free to explore those ideas - I’m not very familiar with IME setups so can’t tell if that approach is feasible or not.

We do have a work-around for basic remapping on the login screen: https://github.com/ililim/dual-key-remap#rebind-keys-during-login--before-windows-loads

**Reply by osteotek** [reply]
Windows Defender finds malware in this release executable (Trojan:Win32/Spursint.F!cl). Can you confirm this as false positive?

**Reply by ililim** [reply]
Yes it’s a false positive. Not sure why it gets flagged - probably because it intercepts keystrokes and it’s not a widely known app.
